### ISSUE TITLE
nix doesn't reset SIGPIPE handler in children processes

### ISSUE THREAD

**Issue by luke-clifton** [NixOS/nix#3022]
Seems that nix doesn't reset the SIGPIPE handler in forked processes.

The following derivation will not build (single user, macOS)

```
let
    nixpkgs = import <nixpkgs> {};
in
    nixpkgs.runCommand "sigpipe-test" {} ''
        yes | true > $out
    ''
```

I would expect this to fail without any output, instead it fails and `yes` prints an error message about a broken pipe because the SIGPIPE didn't kill it.

Possibly related to https://github.com/NixOS/nix/issues/2803 and https://github.com/NixOS/nix/issues/2653

**Comment by luke-clifton** [comment]
```
$ nix --version
nix (Nix) 2.2.2
```

**Comment by luke-clifton** [comment]
Hmm.. this might be a nixpkgs issue.

**Comment by lilyball** [comment]
This is because the generic builder sets `set -s pipefail`, and does not unset it before running the `buildCommand`. It also sets `set -e`, so `false` would cause the command to fail as well.

So this isn't a SIGPIPE handler issue, it's just because `yes` exits with a nonzero status due to `stdout` being closed.

**Comment by luke-clifton** [comment]
You are right, this fails because of `set -o pipefail`, but the reason `yes` is failing is what is actually interesting.

With the default SIGPIPE handler in place, yes terminates with a an error code 141 (sigpipe). However, in that example, `yes` is actually failing with error code 1, and is printing out 

```
yes: standard output: Broken pipe
```

Setting `set -o pipefail` doesn't actually change the SIGPIPE handler, so is not really relevant to the bug at hand.

If you run this in a bash shell

```
set -o pipefail
yes | true
```
You should get no output, however, in the above derivation you get

```
yes: standard output: Broken pipe
```

**Comment by lilyball** [comment]
I can't speak as to the error handler, but my point is, the derivation is expected to fail because `yes | true` results in the build script exiting with a non-zero code due to `set -eo pipefail`.

**Comment by luke-clifton** [comment]
Modified the original comment to make it clearer. Thanks.

**Comment by lilyball** [comment]
FWIW I just tested your original derivation and I see no output and it fails with status 141. This is on macOS.

```
bash-3.2$ nix-build -E $'let
>       nixpkgs = import <nixpkgs> {};
>   in
>       nixpkgs.runCommand "sigpipe-test" {} \'\'
>           yes | true > $out
>       \'\''
these derivations will be built:
  /nix/store/gjd23wbwffrskzifp3qi8zxspyrc2f1l-sigpipe-test.drv
building '/nix/store/gjd23wbwffrskzifp3qi8zxspyrc2f1l-sigpipe-test.drv'...
builder for '/nix/store/gjd23wbwffrskzifp3qi8zxspyrc2f1l-sigpipe-test.drv' failed with exit code 141
error: build of '/nix/store/gjd23wbwffrskzifp3qi8zxspyrc2f1l-sigpipe-test.drv' failed
```

**Comment by luke-clifton** [comment]
Just checked on my Darwin machine.

```
$ uname -sv
Darwin Darwin Kernel Version 16.7.0: Mon Nov 13 21:56:25 PST 2017; root:xnu-3789.72.11~1/RELEASE_X86_64
$ nix --version
nix (Nix) 2.2.2
$ cat test.nix
let
    nixpkgs = import <nixpkgs> {};
in
    nixpkgs.runCommand "sigpipe-test" {} ''
        yes | true > $out
    ''
$ nix build -f test.nix
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
builder for '/nix/store/c37dlm439m9z1fsi4n8mlbcan391gp6b-sigpipe-test.drv' failed with exit code 1; last 1 log lines:
  yes: standard output: Broken pipe
[0 built (1 failed)]
error: build of '/nix/store/c37dlm439m9z1fsi4n8mlbcan391gp6b-sigpipe-test.drv' failed
```

**Comment by lilyball** [comment]
I'm running macOS 10.14.6, maybe there was a behavioral difference with the OS?

**Comment by luke-clifton** [comment]
Ok, my Darwin machine is broken for other reasons (bad terminal emulator it seems), and my linux machine only has the behaviour when using the daemon. So I think this is actually a duplicate of https://github.com/NixOS/nix/issues/2803

Sorry for the noise.

And sorry for dragging you into it @lilyball, I just saw your message about errors on stdout and `yes` while I was on the train and thought it may have been a similar issue, once I got home and read your messages in more detail it was obvious it was probably unrelated.
